src/lib/firewall.ts

```diff
@@ -0,0 +1,198 @@
+import * as _ from 'lodash';
+
+import * as config from '../config/index';
+import * as constants from './constants';
+import * as iptables from './iptables';
+import { log } from './supervisor-console';
+
+import * as dbFormat from '../device-state/db-format';
+
+export const initialised = (async () => {
+	await config.initialized;
+	await applyFirewall();
+
+	// apply firewall whenever relevant config changes occur...
+	config.on('change', async ({ firewallMode, localMode }) => {
+		if (firewallMode || localMode !== undefined) {
+			applyFirewall({ firewallMode, localMode });
+		}
+	});
+})();
+
+const BALENA_FIREWALL_CHAIN = 'BALENA-FIREWALL';
+const LOG_PREFIX = '\u{1F525}';
+
+const prepareChain: iptables.Rule[] = [
+	{
+		action: '-N',
+	},
+	{
+		action: '-F',
+	},
+];
+
+const standardServices: iptables.Rule[] = [
+	{
+		comment: 'SSH Server',
+		action: '-A',
+		proto: 'tcp',
+		matches: ['--dport 22222'],
+		target: 'ACCEPT',
+	},
+	{
+		comment: 'balenaEngine',
+		action: '-A',
+		proto: 'tcp',
+		matches: ['--dport 2375'],
+		target: 'ACCEPT',
+	},
+	{
+		comment: 'mDNS',
+		action: '-A',
+		matches: ['-m addrtype', '--dst-type MULTICAST'],
+		target: 'ACCEPT',
+	},
+	{
+		comment: 'ICMP',
+		action: '-A',
+		proto: 'icmp',
+		target: 'ACCEPT',
+	},
+];
+
+const standardPolicy: iptables.Rule[] = [
+	{
+		comment: 'Locally-sourced traffic',
+		action: '-I',
+		matches: ['-m addrtype', '--src-type LOCAL'],
+		target: 'ACCEPT',
+	},
+	{
+		action: '-I',
+		matches: ['-m state', '--state ESTABLISHED,RELATED'],
+		target: 'ACCEPT',
+	},
+	{
+		comment: 'Reject everything else',
+		action: '-A',
+		target: 'REJECT',
+	},
+];
+
+let supervisorAccessRules: iptables.Rule[] = [];
+function updateSupervisorAccessRules(
+	localMode: boolean,
+	allowedInterfaces: string[],
+	port: number,
+) {
+	supervisorAccessRules = [];
+
+	// if localMode then we're listening on ALL interfaces, otherwise we listen only on the allowed ones...
+	if (localMode) {
+		supervisorAccessRules.push({
+			comment: 'Supervisor API',
+			action: '-A',
+			proto: 'tcp',
+			matches: [`--dport ${port}`],
+			target: 'ACCEPT',
+			applyIfExists: false,
+		});
+	} else {
+		allowedInterfaces.forEach((intf) => {
+			supervisorAccessRules.push({
+				comment: 'Supervisor API',
+				action: '-A',
+				proto: 'tcp',
+				matches: [`--dport ${port}`, `-i ${intf}`],
+				target: 'ACCEPT',
+				applyIfExists: false,
+			});
+		});
+	}
```

Comment:
```suggestion

			supervisorAccessRules.push({
				comment: 'Supervisor API',
				action: '-A',
				proto: 'tcp',
				matches: [`--dport ${port}`].concat(localMode? [] : [`-i ${intf}`']),
				target: 'ACCEPT',
				applyIfExists: false,
			});
	}
```

Reply:
there's also a difference in that the non-localmode path has to create this rule for each interface

Reply:
also, you cannot pass them in one hit -- `multiple -i flags not allowed`